osutil/disks/disks.go

```diff
@@ -0,0 +1,282 @@
+			d.minor = mounts[i].DevMinor
+			partMountPointSource = mounts[i].MountSource
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
+	}
+
+	// TODO:UC20: if the mountpoint is of a decrypted mapper device, then we
+	//            need to trace back from the decrypted mapper device through
+	//            luks to find the real encrypted partition underneath the
+	//            decrypted one and thus the disk device for that partition
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(partMountPointSource)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", partMountPointSource, err)
+	}
+
+	// ID_PART_ENTRY_DISK will give us the major and minor of the disk that this
+	// partition originated from
+	if majorMinor, ok := props["ID_PART_ENTRY_DISK"]; ok {
+		maj, min, err := parseDeviceMajorMinor(majorMinor)
+		if err != nil {
+			// bad udev output?
+			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", partMountPointSource, err)
+		}
+		d.major = maj
+		d.minor = min
+	} else {
+		// the partition is probably a volume or other non-physical disk, so
+		// confirm that DEVTYPE == disk and return the maj/min for it
+		if devType, ok := props["DEVTYPE"]; ok {
+			if devType == "disk" {
+				return d, nil
+			}
+			// unclear what other DEVTYPE's we should support for this
+			return nil, fmt.Errorf("unsupported DEVTYPE %q for mount point source %s", devType, partMountPointSource)
+		}
+
+		return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", partMountPointSource)
+	}
+
+	return d, nil
+}
+
+func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
+	// if we haven't found the partitions for this disk yet, do that now
+	if d.partitions == nil {
+		d.partitions = make(map[string]string)
+		// step 1. find all devices with a matching major number
+		// step 2. start at the major + minor device for the disk, and iterate over
+		//         all devices that have a partition attribute, starting with the
+		//         device with major same as disk and minor equal to disk minor + 1
+		// step 3. if we hit a device that does not have a partition attribute, then
+		//         we hit another disk, and shall stop searching
+
+		// TODO: are there devices that have structures on them that show up as
+		//       contiguous devices but are _not_ partitions, i.e. some littlekernel
+		//       devices?
+
+		// start with the minor + 1, since the major + minor of the disk we have
+		// itself is not a partition
+		currentMinor := d.minor
+		for {
+			currentMinor++
+			partMajMin := fmt.Sprintf("%d:%d", d.major, currentMinor)
```

Comment:
This warrants a comment that even for MBR disks, with logical/extended partition nodes jumping to eg. `/dev/sd*5`, the minor numbers are contiguous.

Reply:
I've looked at bug report: https://bugs.launchpad.net/snappy/+bug/1878374 there's a lsblk output  from that infamous nvme drive in NUC. For some reason the minor numbers are not contiguous:

```
NAME        MAJ:MIN RM         SIZE RO TYPE MOUNTPOINT
nvme0n1     259:0    0 128035676160  0 disk 
|-nvme0n1p2 259:2    0   1258291200  0 part /writable/system-data/var/lib/snapd/seed
|-nvme0n1p1 259:3    0      1048576  0 part 
|-nvme0n1p3 259:4    0    786432000  0 part 
`-nvme0n1p4 259:5    0 125988838912  0 part 
```

However, they are definitely contiguous in my system:

```
nvme0n1                                       259:0    0 232.9G  0 disk  
├─nvme0n1p1                                   259:1    0   512M  0 part  /boot/efi
├─nvme0n1p2                                   259:2    0     1G  0 part  /boot
└─nvme0n1p3                                   259:3    0 231.4G  0 part  
  └─luks-3934e9a7-4e21-4dde-a52e-d01055a92a83 254:0    0 231.4G  0 crypt 
    ├─linux-root                              254:1    0    30G  0 lvm   /
    └─linux-home                              254:2    0 201.4G  0 lvm   /home
```

Reply:
fun, do we know if 259:1 doesn't exist at all on that system?  or it exists but is something strange that lsblk ignores/skips?

Reply:
I will inquire about this in the bug report, but it will be at least another day before we get an answer.

Reply:
as per the final comments in that bug report, there is a contiguous major/minor number for all block devices on a successful install on this NVMe, so likely this was an artifact of the bug with the intel SSD firmware